### PR TITLE
fix for windows debug support

### PR DESCRIPTION
**Pull request by snoglobe**
### What does this PR do?

this fixes https://github.com/oven-sh/bun/issues/8851, which pertains to the bun debugger not working on windows.

for windows platform, we now use websockets to communicate to the debugger and inspector. on unix platforms, we keep the existing unix sockets implementation.

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

tested it by debugging programs under vscode in windows

### CODE REVIEW

**Comment by robobun** [comment]
@snoglobe, your commit bc666970122eb45c7d090b462828547b7bb3f8f6 is building: [`#3614`](https://buildkite.com/bun/bun/builds/3614)

**Comment by snoglobe** [comment]
note: failed builds are due to buildkite

File: packages/bun-debug-adapter-protocol/src/debugger/adapter.ts [hunk 1]
```diff
@@ -2,12 +2,19 @@ import type { InspectorEventMap } from "../../../bun-inspector-protocol/src/insp
...
-import { spawn } from "node:child_process";
+import { spawn, ChildProcess } from "node:child_process";
 import { EventEmitter } from "node:events";
 import { WebSocketInspector, remoteObjectToString } from "../../../bun-inspector-protocol/index";
-import { UnixSignal, randomUnixPath } from "./signal";
+import { randomUnixPath, UnixSignal, WebSocketSignal } from "./signal";
 import { Location, SourceMap } from "./sourcemap";
+import { createServer, AddressInfo } from "node:net";
+
+export function getAvailablePort(): number {
```
Comment (Jarred-Sumner) [inline packages/bun-debug-adapter-protocol/src/debugger/adapter.ts]: Do we need to wait for the callback to be called?

File: packages/bun-vscode/scripts/build.mjs [hunk 2]
```diff
@@ -9,7 +9,7 @@ buildSync({
   entryPoints: ["src/extension.ts", "src/web-extension.ts"],
   outdir: "dist",
   bundle: true,
-  external: ["vscode"],
+  external: ["vscode", "ws"],
```
Comment (Jarred-Sumner) [inline packages/bun-vscode/scripts/build.mjs]: Is this intentional?

**Comment by snoglobe** [comment]
yes; build fails otherwise

**Comment by Jarred-Sumner** [comment]
What happens when you build the installed extension and the ws package
isn’t installed?

File: packages/bun-debug-adapter-protocol/src/debugger/adapter.ts (hunk 1 above)
Comment (snoglobe) [reply packages/bun-debug-adapter-protocol/src/debugger/adapter.ts]: it works fine without waiting for callback

**Comment by Jarred-Sumner** [comment]
> errors w/ the bundler unable to resolve ws

did you run `bun install` (or npm install)?

File: packages/bun-debug-adapter-protocol/src/debugger/adapter.ts (hunk 1 above)
Comment (Jarred-Sumner) [reply packages/bun-debug-adapter-protocol/src/debugger/adapter.ts]: does node gurantee that the address will always be set before the listen callback is invoked?

File: src/js/internal/debugger.ts [hunk 3]
```diff
@@ -73,7 +83,7 @@ class Debugger {
   #listen(): void {
     const { protocol, hostname, port, pathname } = this.#url;
 
-    if (protocol === "ws:" || protocol === "ws+tcp:") {
```
Comment (Jarred-Sumner) [inline src/js/internal/debugger.ts]: If we're allowing wss, where is the TLS configuration specified?

File: packages/bun-debug-adapter-protocol/src/debugger/adapter.ts (hunk 1 above)
Comment (Jarred-Sumner) [reply packages/bun-debug-adapter-protocol/src/debugger/adapter.ts]: > server.address() returns null before the 'listening' event has been emitted or after calling server.close().

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
Not mergeable until:
1) the `listen()` callback is addressed (it will be randomly flaky, in it's current version even if it appears to work sometimes when you run it locally)
2) The `"ws"` module should not be external. We bundle the VSCode extension instead of expecting a node_modules folder to be present (it won't be, except for development) 

Separately, the `"wss"` validation is a little confusing unless the idea is you have a reverse proxy between Bun.serve() and the websocket connection? This is uncommon, as people usually don't proxy websockets but not impossible I suppose.

File: src/js/internal/debugger.ts [hunk 4]
```diff
@@ -321,6 +331,18 @@ function reset(): string {
   return "";
 }
 
+function notifyWebSocket(url: string): void {
```
Comment (Jarred-Sumner) [inline src/js/internal/debugger.ts]: Oh, it's a whole websocket server for this? We can just make a TCP connection.

**Comment by Jarred-Sumner** [comment]
I want to get the Bun changes into the release, so going to merge that separately from this PR since the extension can be updated separately after

**Comment by snoglobe** [comment]
> Not mergeable until:
> 
> 1. the `listen()` callback is addressed (it will be randomly flaky, in it's current version even if it appears to work sometimes when you run it locally)
> 2. The `"ws"` module should not be external. We bundle the VSCode extension instead of expecting a node_modules folder to be present (it won't be, except for development)
> 
> Separately, the `"wss"` validation is a little confusing unless the idea is you have a reverse proxy between Bun.serve() and the websocket connection? This is uncommon, as people usually don't proxy websockets but not impossible I suppose.

will remove wss validation then; yeah not sure why i put it

**Comment by snoglobe** [comment]
> > errors w/ the bundler unable to resolve ws
> 
> did you run `bun install` (or npm install)?

yes

File: packages/bun-debug-adapter-protocol/src/debugger/adapter.ts [hunk 5]
```diff
@@ -2149,10 +2207,10 @@ function titleize(name: string): string {
 }
 
 function sourcePresentationHint(url?: string): DAP.Source["presentationHint"] {
-  if (!url || !url.startsWith("/")) {
+  if (!url || path.isAbsolute(url)) {
```
Comment (Jarred-Sumner) [inline packages/bun-debug-adapter-protocol/src/debugger/adapter.ts]: !path.isAbsolute

File: packages/bun-debug-adapter-protocol/src/debugger/adapter.ts [hunk 6]
```diff
@@ -2567,3 +2628,11 @@ let sequence = 1;
 function nextId(): number {
   return sequence++;
 }
+
+export function getRandomId() {
+  return Math.random().toString(36).slice(2);
+}
+
+export function normalizeWindowsPath(path: string): string {
+  return (path.charAt(0).toUpperCase() + path.slice(1)).replaceAll("\\\\", "\\");
```
Comment (Jarred-Sumner) [inline packages/bun-debug-adapter-protocol/src/debugger/adapter.ts]:
```suggestion
  path = normalize(path);
  if (path[1] === ':' && (path[2] === "\\" || path[2] === "/")) { 
    return (path.charAt(0).toUpperCase() + path.slice(1)).replaceAll("\\\\", "\\");
  }
   return path;
```

File: packages/bun-debug-adapter-protocol/src/debugger/adapter.ts [hunk 7]
```diff
@@ -2163,6 +2221,9 @@ function sourceName(url?: string): string {
     return "unknown.js";
   }
   if (isJavaScript(url)) {
+    if (process.platform === "win32") {
+      return url.split("\\").pop() || url;
```
Comment (Jarred-Sumner) [inline packages/bun-debug-adapter-protocol/src/debugger/adapter.ts]:
```suggestion
      url = url.replaceAll("\\", "/");
```

File: packages/bun-vscode/src/features/debug.ts [hunk 8]
```diff
@@ -211,18 +215,18 @@ class TerminalDebugSession extends FileDebugSession {
...
-      this.signal.on("Signal.received", () => {
-        vscode.debug.startDebugging(undefined, {
-          ...ATTACH_CONFIGURATION,
-          url: this.adapter.url,
-        });
+      });
+    } else {
+      this.signal = new UnixSignal();
+    }
+    this.signal.on("Signal.received", () => {
```
Comment (Jarred-Sumner) [inline packages/bun-vscode/src/features/debug.ts:234]: signal is undefined here

File: packages/bun-vscode/src/features/debug.ts [hunk 9]
```diff
@@ -176,19 +176,23 @@ class FileDebugSession extends DebugSession {
...
+    let url;
+    if (process.platform === "win32") {
+      getAvailablePort().then(port => {
+        // the signal is only first used well after this resolves
+        url = `ws://127.0.0.1:${port}/${getRandomId()}`;
+      });
+    } else {
+      url = `ws+unix://${tmpdir()}/${uniqueId}.sock`;
+    }
+    this.adapter = new DebugAdapter(url);
```
Comment (Jarred-Sumner) [inline packages/bun-vscode/src/features/debug.ts:195]: url is undefined here

File: packages/bun-debug-adapter-protocol/src/debugger/signal.ts [hunk 10]
```diff
@@ -85,3 +85,72 @@ function parseUnixPath(path: string | URL): string {
...
+  #server: ReturnType<typeof createServer>;
+  #ready: Promise<void>;
+
+  constructor(port: number) {
+    super();
+    this.#port = port;
+
+    this.#server = createServer((socket: Socket) => {
+      socket.on('data', (data) => {
+        console.error('received', data);
```
Comment (Jarred-Sumner) [inline packages/bun-debug-adapter-protocol/src/debugger/signal.ts]:
```suggestion

```

File: packages/bun-debug-adapter-protocol/src/debugger/adapter.ts [hunk 11]
```diff
@@ -489,36 +503,73 @@ export class DebugAdapter extends EventEmitter<DebugAdapterEventMap> implements
...
 
-    if (!started) {
-      throw new Error("Program could not be started.");
+      this.once("Adapter.terminated", () => {
+        signal.close();
+      });
+
+      const query = stopOnEntry ? "break=1" : "wait=1";
+      processEnv["BUN_INSPECT"] = `${url}?${query}`;
+      processEnv["BUN_INSPECT_NOTIFY"] = `tcp://127.0.0.1:${signal.port}`; // 127.0.0.1 so it resolves correctly on windows
```
Comment (Jarred-Sumner) [inline packages/bun-debug-adapter-protocol/src/debugger/adapter.ts]:
```suggestion
      processEnv["BUN_INSPECT_NOTIFY"] = signal.url
```

**Comment by liudonghua123** [comment]
I see this pr changed unix domain socket to websocket through network ipc. It just like bun inspect cli debugging doing.

I tested some simple demos of unix domain socket on windows and it works. How about to fix existing unix domain socket for windows, it's more high performance in my opinion. It's just my own thought. :smile:

**Comment by Jarred-Sumner** [comment]
@liudonghua123 you're welcome to try getting unix domain sockets working when used in VSCode on Windows

**Pull request merged**